includes/class-wcpdf-main.php

```diff
@@ -131,70 +143,80 @@ public function attach_document_to_email( $attachments, $email_id, $order, $emai
 		}
 
 		$attach_to_document_types = $this->get_documents_for_email( $email_id, $order );
+		$lock_name                = $this->lock_name . '_attach_document_to_email_' . $order_id;
```

Comment:
Does it also need to contain the document type and maybe the output type?

Reply:
I added the order number to avoid having concurrent orders being processed at the same time, for example, if 2 customers are buying at the exact same time.

I don't think that the document type or the output type helps here.

Reply:
> Also, Is it better to put the if above the foreach?

Yes, maybe it's better.

Reply:
Regarding the first comment [here](https://wordpress.org/support/topic/missing-4-invoices-95-99-102-and-139/), I think we should avoid that.

Reply:
Can you help me understand something here? It seems a little confusing to me.

Consider, for example, ten requests at the same time (that calls the `attach_document_to_email` method ten times). One of them passes through the `if ( $lock->lock( $this->lock_retries ) ) {`, for the other nine, since the request won't freeze here until the lock gets released, then all of them fail. Am I right?

Reply:
Well, only if there are many at the same time and the process takes too much, which seems unlikely no?

Reply:
> which seems unlikely no?

Yes, I think it hardly can happen. Maybe only for big stores with lots of orders, and still unlikely.
Thanks for the explanation.

Reply:
Yes, my idea is avoiding having 2 or more orders processing the attachments at the same time to avoid concurrent document creations, which should avoid issues with the numbers and assignments.